lib/rubygems/specification_policy.rb

```diff
@@ -120,6 +120,19 @@ def validate_optional(strict)
     end
   end
 
+  ##
+  # Implementation for Specification#validate_for_resolution
+
+  def validate_for_resolution
+    validate_required_attributes
+
+    validate_name
+
+    validate_platform
+
+    validate_duplicate_dependencies
+  end
```

Comment:
Is there anything else that should be validated for resolution?

Reply:
`validate` is also called [on the rubygems.org side](https://github.com/rubygems/rubygems.org/blob/1420624b01bbd14b32059831c2b95a8713dbb754/app/models/pusher.rb#L292-L294) with `packaging = false` (and also `strict = false`), so if metadata validations were dependent on `packaging` flag, they would no longer be performed in that context. Is that correct?

Reply:
Great catch @thomasmarshall! I think that's the behavior we definitely don't want to change, since that's what actually prevents invalid gems from being pushed (you can easily build an invalid gem in the client side, with `--force` flag for example).

It seems that was introduced not long ago to fix a [security issue](https://github.com/rubygems/rubygems.org/commit/7e19c19247ddf5885a915710afc60ec6663d8502), but I guess it's also having the side effect of validating other things not involved with the security issue.

One if the errors we try to avoid here seems to be actually skipped there too, by monkey
patching `Gem::SpecificationPolicy`: https://github.com/rubygems/rubygems.org/blob/7e19c19247ddf5885a915710afc60ec6663d8502/app/models/specification_policy.rb, so if we went with modifying packaging we would not change the behavior there and would allow the monkeypatch to be removed. However, there's also an "invalid link"  error that I think would be lost.

So, I'm not sure where to go from here, opinions?

Reply:
Hmm how about a flag which skips validations for fields matching `LAZY_PATTERN` like `ignore_todos`? https://github.com/rubygems/rubygems/blob/a15e2cb599e1415d0e9577eb7ca2d0df985a4c38/lib/rubygems/specification_policy.rb#L416

Reply:
The problem is we not only need to skip "TODO checks" but also "invalid link" checks. I think those would be the minimal errors that need to be skipped. Directly changing the  `packaging` to skip TODOs by default seems like it would be fine, since that's what `rubygems.org` is indirectly doing anythings. But changing `packaging` to skip invalid links would change behavior. So we could add a special "skip_invalid_links" flag, but feels arbitrary to me. The initial approach here of a `validate_for_resolution` method seems better to me, and I'd probably just make it so that it does everything done today, except for TODOs and invalid links. Alternatively we could change to `packaging` flag to a `phase` flag that could be ":resolution` (new mode for Bundler), `:build` (default mode), or `:push` (old `packaging == true` for `rubygems.org`), but maybe that's overkill and more complicated.

Reply:
OK cool, I pushed up a new version which adds the monkeypatch for `validate_for_resolution` and sets the logic to validate everything required except metadata fields.

The `phase` idea sounds good too, I'm happy to give it a go if you think that's a cleaner solution in the long run.

Reply:
I think the current approach is good enough, and I'm not sure about adding a `phase` parameter so let's move on!